### PR TITLE
Add a FilterEmptyDatasetsTool and allow filtering tools to apply to arbitrary collection structures

### PR DESCRIPTION
**Pull request by alimatai**
Hello,

Here is an extra-condition for the tool ``FilterFailedDatasetsTool`` to allow it to remove empty datasets as well as failed datasets in a collection. I tested it on my data and it seems to work fine. Since it's my first time looking at galaxy's source code, I don't know if there is a proper way to do that ... And I didn't make the improvement for paired datasets because I don't have proper data to test it. But I assume it would be something like ``if forward_o.is_ok and reverse_o.is_ok and forward_o.has_data() and reverse_o.has_data():`` ?

### CODE REVIEW

**Comment by mvdbeek** [comment]
Hey @Mataivic, this is an interesting improvement, but I think this behaviour should be selectable in the tool form, since empty datasets are usually not failed datasets (if they are failed that should be configured in the tool that produced the output). You can define additional parameters in the xml of this tool (https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/tools/filter_failed_collection.xm), just like you'd do for regular tools. The parameters are then available in `incoming`, which is a simple dictionary, see for example https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/tools/__init__.py#L2567 how to use this.

**Comment by alimatai** [comment]
@mvdbeek It may not be the most elegant way to do, but I applied your suggestion. I also made the improvement for paired dataset but I have no idea if it works the same way than regular collections (I guess the Travis test will tell us that) ; feel free to cancel that one or to modify it.

**Comment by alimatai** [comment]
@mvdbeek Hello, Just remembered I made this PR ... Is is acceptable ?

**Comment by mvdbeek** [comment]
Sorry, forgot about this, I do have a couple of changes that I'll try to get in.

**Comment by mvdbeek** [comment]
OK, so this took far longer than I had hoped, sorry for holding this up so long.
Ultimately I think it's better to have a separate tools for filtering failed and empty datasets from a collection, so I've added 2 commits. This should now work for any type of collection structure, not just list, paired or list:paired.

**Comment by mvdbeek** [comment]
@galaxybot test this

File: lib/galaxy/tools/filter_empty_collection.xml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+                  <assert_contents>
+                      <has_text_matching expression="^This is a line of text.\n$" />
+                  </assert_contents>
+                </element>
+            </output_collection>
+        </test>
+    </tests>
+    <help><![CDATA[
+
+This tool takes a list dataset collection and filters out empty datasets. This is useful for continuing a multi-sample analysis when downstream tools require datasets to have content.
```
Comment (nsoranzo) [inline lib/galaxy/tools/filter_empty_collection.xml]: `s/list dataset collection/dataset collection/`

File: lib/galaxy/tools/__init__.py [hunk 2]
```diff
@@ -2510,42 +2510,64 @@ def produce_outputs(self, trans, out_data, output_collections, incoming, history
         )
 
 
-class FilterFailedDatasetsTool(DatabaseOperationTool):
-    tool_type = 'filter_failed_datasets_collection'
-    require_dataset_ok = False
+class FilterDatasetsTool(DatabaseOperationTool):
+
+    def get_new_elements(self, history, elements_to_copy):
```
Comment (nsoranzo) [inline lib/galaxy/tools/__init__.py]: Maybe make it "private" by renaming it to `_get_new_elements` ?

Also, maybe swap the `history` and `elements_to_copy` parameters in line with how the method is called below?

File: lib/galaxy/tools/__init__.py [hunk 3]
```diff
@@ -2510,42 +2510,64 @@ def produce_outputs(self, trans, out_data, output_collections, incoming, history
...
+        # We only process list or paired collections. Higher order collection will be mapped over
+        assert collection_type in ("list", "paired")
+
+        elements_to_copy = []
+        for element in elements:
+            if self.element_is_valid(element):
+                elements_to_copy.append(element)
+            elif collection_type == 'paired':
+                # One of the pairs is not OK, skip creating output for this pair
+                return
```
Comment (nsoranzo) [inline lib/galaxy/tools/__init__.py]: Is it OK to just `return` here or should this return e.g. `self._outputs_dict()` as `DatabaseOperationTool.produce_outputs()` does?

Comment (mvdbeek) [reply lib/galaxy/tools/__init__.py]: Good point, it actually isn't

File: lib/galaxy/tools/__init__.py (hunk 2 above)
Comment (mvdbeek) [reply lib/galaxy/tools/__init__.py]: I've flipped the caller methods, that seems more "natural" to me and that's how we handle a lot of the signatures in the test framework

**Comment by mvdbeek** [comment]
@galaxybot test this

**Comment by mvdbeek** [comment]
@galaxybot test this

File: lib/galaxy/tools/__init__.py [hunk 4]
```diff
@@ -2510,42 +2510,68 @@ def produce_outputs(self, trans, out_data, output_collections, incoming, history
...
-        assert hdca.collection.collection_type == "list" or hdca.collection.collection_type == 'list:paired'
+        if hasattr(collection, 'element_object'):
+            # A list
+            elements = collection.element_object.elements
+            collection_type = collection.element_object.collection_type
+        else:
+            # A pair
+            elements = collection.collection.elements
+            collection_type = collection.collection.collection_type
+        # We only process list or paired collections. Higher order collection will be mapped over
```
Comment (nsoranzo) [inline lib/galaxy/tools/__init__.py]: Should lines 2533 and 2536 be updated to say `list of pair` instead of `pair`/`paired collections` ?

**Review by nsoranzo — APPROVED** [review]

**Comment by mvdbeek** [comment]
Thanks for putting this up here @Mataivic, very useful functionality!

**Comment by alimatai** [comment]
@mvdbeek You're welcome :). Thanks for your help !

**Pull request merged**